Objective-C/CBLReplicator.mm

```diff
@@ -165,17 +165,53 @@ - (void) start {
     }
 }
 
-- (void) retry: (BOOL)reset {
-    CBL_LOCK(self) {
-        CBLLogInfo(Sync, @"%@: Retrying...", self);
-        if (reset)
-            _retryCount = 0;
-        if (_repl || _state <= kCBLStateStopping) {
-            CBLLogInfo(Sync, @"%@: Ignore retrying (state = %d, status = %d)",
-                       self, _state, _rawStatus.level);
-            return;
+- (void) resetAndScheduleRetry {
+    [self resetRetryCount];
+    [self scheduleRetry: 0];
+}
+
+- (void) scheduleRetry: (NSTimeInterval)delay {
+    [self cancelPrviousRetry];
+    
+    // to register the perform selector in the main run loop
+    dispatch_async(dispatch_get_main_queue(), ^{
+        [self performSelector: @selector(retry)
+                   withObject: nil
+                   afterDelay: delay];
+    });
+}
+
+- (void) resetRetryCount {
+    CBLLogVerbose(Sync, @"%@: Resetting the retry count", self);
+    _retryCount = 0;
+}
+
+- (void) cancelPrviousRetry {
+    __weak CBLReplicator *weakSelf = self;
+    // remove already registered perform selector(retry) from main run loop
+    dispatch_async(dispatch_get_main_queue(), ^{
+        CBLReplicator *strongSelf = weakSelf;
+        if (strongSelf) {
+            [NSObject cancelPreviousPerformRequestsWithTarget: strongSelf
+                                                     selector: @selector(retry)
+                                                       object: nil];
         }
-        [self _start];
+    });
+}
+
+- (void) retry {
+    if (_state == kCBLStateOffline || _state == kCBLStateSuspended) {
+        dispatch_async(_dispatchQueue, ^{
```

Comment:
Why do we need to call in the dispatch queue?

Reply:
else this will be called from main queue